libs/langgraph/src/langgraph-architecture-spec.ts

```diff
@@ -0,0 +1,399 @@
+ * ====================================
+ *
+ * The Checkpointer layer provides persistence capabilities, allowing the system
+ * to save, restore, and replay execution states.
+ *
+ * Key Interfaces & Classes:
+ * - BaseStore: Abstract interface for storing and retrieving checkpoints
+ * - BaseCheckpointSaver: Abstract interface for saving checkpoints
+ * - MemoryStore: In-memory implementation of BaseStore
+ * - CheckpointTuple: Represents a checkpoint with ID, state and metadata
+ * - PendingWrite: Represents a write operation that can be batched
+ *
+ * Plugin Implementations:
+ * - SQLiteStore: Persists checkpoints to SQLite database
+ * - PostgresStore: Persists checkpoints to PostgreSQL database
+ * - MongoDBStore: Persists checkpoints to MongoDB database
+ *
+ * Core Functionality:
+ * - Snapshot Creation: Capture complete system state at a point in time
+ * - Storage & Retrieval: Save and load checkpoints from different backends
+ * - Namespace Management: Organize checkpoints in hierarchical namespaces
+ * - Versioning: Track and manage checkpoint versions
+ * - Time Travel: Enable replay and debugging from past states
+ */
+
+/**
+ * =======================================
+ * Layer 3: Pregel - Execution Engine
+ * =======================================
+ *
+ * The Pregel layer is the core execution engine implementing the message-passing
+ * computational model.
+ *
+ * Key Classes:
+ * - Pregel: Core runtime engine
+ * - PregelNode: Represents a computation unit that subscribes to channels
+ * - ChannelWrite: Specifies how nodes write to channels
+ * - PregelLoop: Controls the execution flow through supersteps
+ * - PregelRunner: Manages concurrent execution of tasks
+ *
+ * Core Functionality:
+ * - Graph Execution: Orchestrate message flow between nodes
+ * - State Propagation: Update and propagate state throughout graph
+ * - Concurrency: Execute nodes in parallel when possible
+ * - Error Handling: Manage failures and retries
+ * - Streaming: Provide real-time visibility into execution progress
+ * - Persistence: Save and restore execution state
+ */
+
+/**
+ * =======================================
+ * Layer 4: Graph - Workflow Definition
+ * =======================================
+ *
+ * The Graph layer provides high-level interfaces for building workflows
+ * that leverage the underlying Pregel system.
+ *
+ * Key Classes:
+ * - Graph: Base graph class for defining workflow structure
+ * - StateGraph: Specialized graph for stateful workflows
+ * - CompiledGraph (extends Pregel): Runtime representation of a graph
+ * - Annotation: Type specification for graph state
+ * - Branch: Handles conditional routing between nodes
+ *
+ * Core Functionality:
+ * - Graph Construction: Define nodes, edges, and execution flow
+ * - State Definition: Specify and validate state schemas
+ * - Compilation: Transform graph definitions into executable Pregel instances
+ * - Visualization: Generate visualizations of graph structure
```

Comment:
Might be hard to define here, but it'd be nice if there was some (even terse) statements that said how each of these map to the pregel model